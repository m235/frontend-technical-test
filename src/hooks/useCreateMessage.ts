import { AxiosError } from 'axios'
import { useMutation } from 'react-query'
import { UseMutationOptions } from 'react-query/types/react/types'

import type { Message } from '@/types/message'
import { createMessage, MessagesApi } from '@/services/messages'

type PartialMessage = Omit<Message, 'id'>
type Options = UseMutationOptions<Message, AxiosError, PartialMessage>

const useCreateMessage = (conversationId: number, options?: Options) => {
  return useMutation<Message, AxiosError, PartialMessage>([MessagesApi.MESSAGES, conversationId], createMessage, options)
}

export default useCreateMessage
