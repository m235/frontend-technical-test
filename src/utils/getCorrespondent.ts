import type { Conversation } from '@/types/conversation'

export const getCorrespondent = (userId: number, conversation: Conversation) => {
  switch (userId) {
    case conversation.senderId:
      return conversation.recipientNickname
    case conversation.recipientId:
      return conversation.senderNickname
    default:
      throw new Error('You are nor the recipient nor the sender')
  }
}
