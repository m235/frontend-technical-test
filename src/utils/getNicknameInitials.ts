export const getNicknameInitials = (nickname: string) =>
  nickname
    .split(' ')
    .map((string) => string.trim().charAt(0))
    .slice(0, 2)
    .join('')
