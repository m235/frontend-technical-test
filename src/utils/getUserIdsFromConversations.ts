import type { Conversation } from '@/types/conversation'

export const getUserIdsFromConversations = (conversations?: Conversation[]) =>
  Array.from<number>(new Set(conversations?.reduce((acc, curr) => [...acc, curr.recipientId, curr.senderId], []) ?? []))
