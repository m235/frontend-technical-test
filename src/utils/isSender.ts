import type { Conversation } from '@/types/conversation'

export const isSender = (userId: number, conversation: Conversation) => userId === conversation.senderId
