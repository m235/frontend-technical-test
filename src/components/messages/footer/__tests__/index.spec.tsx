import { fireEvent, waitFor } from '@testing-library/dom'

import MessagesFooter from '@/components/messages/footer'
import useAuth from '@/hooks/useAuth'
import useCreateMessage from '@/hooks/useCreateMessage'
import { render } from '@/tests/utils'

const mockedUseAuth = useAuth as jest.Mock<any>
const mockedUseCreateMessage = useCreateMessage as jest.Mock<any>
jest.mock('@/hooks/useAuth')
jest.mock('@/hooks/useCreateMessage')

describe('components:conversations:footer', () => {
  beforeEach(() => {
    mockedUseAuth.mockImplementation(() => ({ userId: 1, user: [{ id: 1, nickname: 'John Travolta', token: 'x' }] }))
    mockedUseCreateMessage.mockImplementation(() => ({ isLoading: false, mutate: () => null }))
  })
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('should render default correctly', () => {
    const { container } = render(<MessagesFooter conversationId={1} />)

    expect(container).toMatchSnapshot()
  })
  it('should have a disabled button by default', () => {
    const { getByTestId } = render(<MessagesFooter conversationId={1} />)

    expect(getByTestId('send-msg-btn')).toBeDisabled()
  })
  it('should send a message if btn active and clicked', async () => {
    mockedUseCreateMessage.mockImplementation(() => ({ isLoading: false, mutate: createFn }))
    const createFn = jest.fn()
    const { getByTestId } = render(<MessagesFooter conversationId={1} />)

    fireEvent.change(getByTestId('input-text'), { target: { value: 'Message' } })

    expect(getByTestId('send-msg-btn')).toBeEnabled()

    fireEvent.click(getByTestId('send-msg-btn'))

    expect(createFn).toHaveBeenCalledTimes(1)
    expect(createFn).toHaveBeenCalledWith(expect.objectContaining({ body: 'Message', authorId: 1, conversationId: 1 }))
  })
})
