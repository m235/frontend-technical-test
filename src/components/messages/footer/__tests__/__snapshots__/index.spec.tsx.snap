// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components:conversations:footer should render default correctly 1`] = `
<div>
  <footer
    class="sc-dkzDqf dUruFM"
  >
    <textarea
      class="sc-gsnTZi efMEkn"
      data-testid="input-text"
      placeholder="Type your message..."
      rows="1"
    />
    <button
      aria-label="Send"
      class="sc-bczRLJ ekYIRc"
      data-testid="send-msg-btn"
      disabled=""
    >
      <i
        class="fa-solid fa-paper-plane"
      />
    </button>
  </footer>
</div>
`;
